distchan.go

```diff
@@ -386,64 +429,86 @@ func (c *Client) handleIncomingMessages() {
 			panic(err)
 		}
 
-		buf.Reset()
-		c.inv.Send(x.Elem())
+		reflect.ValueOf(c.inv).Send(x.Elem())
+	}
+
+	if buf.Cap() <= 1<<22 {
+		putBuffer(buf)
 	}
 }
 
 func (c *Client) handleOutgoingMessages() {
 	var (
-		buf bytes.Buffer
+		buf = getBuffer()
 		enc = gob.NewEncoder(&buf)
 	)
 
 	for {
-		x, ok := c.outv.Recv()
+		buf.Reset()
+
+		x, ok := reflect.ValueOf(c.outv).Recv()
 		if !ok {
 			break
 		}
 		if err := enc.EncodeValue(x); err != nil {
 			c.logger.Panicln(err)
 		}
 
-		b := buf.Bytes()
-		buf.Reset()
-
 		for _, encoder := range c.encoders {
-			b = encoder(b)
+			ec := encoder(&buf)
+			buf.Reset()
+			if _, err := io.Copy(&buf, ec); err != nil {
+				c.logger.Panicln(err)
+			}
 		}
 
-		if err := writeChunk(c.conn, b); err != nil {
+		if err := writeChunk(c.conn, &buf, buf.Len()); err != nil {
 			c.logger.Printf("error writing value to connection: %s\n", err)
 		}
 	}
 
+	if buf.Cap() <= 1<<22 {
+		putBuffer(buf)
+	}
+
 	close(c.done)
 }
 
-func readChunk(r io.Reader) ([]byte, error) {
+var signature int32 = 0x7f38b034 // protecting from bad incoming data
```

Comment:
So the signature is just a quick way to ensure that data isn't being corrupted in transit, I assume? How did you come up with the value `0x7f38b034`?

Reply:
yes, tis is random number, but protect from HTTP traffic, by example

Reply:
I may keep this concept around, but I'm thinking it would be a good idea to make the variable public so that users can change it to a custom value if they want to.

Reply:
ok